File: src/com/JavaSnakes/panels/MenuPanel.java
package com.JavaSnakes.panels;

import com.JavaSnakes.Main;
import com.JavaSnakes.util.MenuCard;

import java.awt.CardLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.Insets;
import java.awt.event.FocusEvent;
import java.awt.event.FocusListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.util.ArrayList;
import java.util.stream.IntStream;
import java.util.List;
import javax.swing.JButton;
import javax.swing.JCheckBox;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JSeparator;
import javax.swing.JSpinner;
import javax.swing.JTextField;
import javax.swing.SpinnerNumberModel;
import javax.swing.SwingConstants;

public class MenuPanel {

    private final Main owner;
    public JPanel mainPanel;

    private CardLayout cardLayout;

    private JButton newGameButton;
    private JButton settingsButton;
    private JButton quitButton;

    private JButton startGameButton;
    private JSpinner playerSnakeSpinner;
    private JSpinner botSnakeSpinner;
    private JCheckBox wallCheckBox;
    private JButton toMainButton1;

    private JButton toMainButton2;
    private JSpinner mapWidthSpinner;
    private JSpinner mapHeightSpinner;
    private JSpinner frameDelaySpinner;
    private JSpinner snakeNumSpinner;
    private JTextField[] dirCtrlFields; // Ordered by up, down, left, and right

    private int maxSnakeCount;
    private List<int[]> playerControls;
    private int[] playerControlsTemp;

    public MenuPanel(Main setOwner) {
        owner = setOwner;

        mainPanel = new JPanel();
        mainPanel.setLayout(new CardLayout());

        maxSnakeCount = 8;
        playerControls = new ArrayList<>();
        playerControls.add(new int[]{KeyEvent.VK_UP, KeyEvent.VK_DOWN, KeyEvent.VK_LEFT, KeyEvent.VK_RIGHT});
        playerControls.add(new int[]{KeyEvent.VK_W, KeyEvent.VK_S, KeyEvent.VK_A, KeyEvent.VK_D});
        playerControlsTemp = new int[4];

        createMainMenuCard();
        createNewGameCard();
        createSettingsCard();

        cardLayout = (CardLayout) mainPanel.getLayout();
        mainPanel.setPreferredSize(new Dimension(300, 300));
    }

    private void createMainMenuCard() {
        newGameButton = new JButton("New Game");
        newGameButton.addActionListener(e -> toNewGameCard());
        settingsButton = new JButton("Settings");
        settingsButton.addActionListener(e -> toSettingsCard());
        quitButton = new JButton("Quit");
        quitButton.addActionListener(e -> System.exit(0));

        MenuCard gridBag = new MenuCard();
        gridBag.addInGrid(newGameButton, 0, 0);
        gridBag.addInGrid(settingsButton, 1, 0);
        gridBag.addInGrid(quitButton, 2, 0);

        mainPanel.add(gridBag, "main menu card");
    }

    private void createNewGameCard() {
        startGameButton = new JButton("Start Game");
        startGameButton.addActionListener(e -> startGame());
        JLabel playerSnakeCountLabel = new JLabel("Player count:");
        playerSnakeSpinner = new JSpinner(new SpinnerNumberModel(1, 0, maxSnakeCount, 1));
        playerSnakeSpinner.addChangeListener(e -> playerSnakeCountChanged());
        JLabel botSnakeCountLabel = new JLabel("Bot count:");
        botSnakeSpinner = new JSpinner(new SpinnerNumberModel(0, 0, maxSnakeCount, 1));
        botSnakeSpinner.addChangeListener(e -> botSnakeCountChanged());
        wallCheckBox = new JCheckBox("Include walls");
        toMainButton1 = new JButton("Back");
        toMainButton1.addActionListener(e -> toMainMenuCard());

        MenuCard gridBag = new MenuCard();
        gridBag.addInGrid(startGameButton, 0, 0, 2);
        gridBag.addInGrid(playerSnakeCountLabel, 1, 0);
        gridBag.addInGrid(playerSnakeSpinner, 1, 1, 1, new Insets(0, 5, 0, 0));
        gridBag.addInGrid(botSnakeCountLabel, 2, 0);
        gridBag.addInGrid(botSnakeSpinner, 2, 1, 1, new Insets(0, 5, 0, 0));
        gridBag.addInGrid(wallCheckBox, 3, 0, 2);
        gridBag.addInGrid(toMainButton1, 4, 0, 2);

        mainPanel.add(gridBag, "new game card");
    }

    private void createSettingsCard() {
        toMainButton2 = new JButton("Back");
        toMainButton2.addActionListener(e -> toMainMenuCard());
        JLabel mapWidthLabel = new JLabel("Map width:");
        mapWidthSpinner = new JSpinner(new SpinnerNumberModel(30, 20, 100, 1));
        JLabel mapHeightLabel = new JLabel("Map height:");
        mapHeightSpinner = new JSpinner(new SpinnerNumberModel(20, 20, 100, 1));
        JLabel frameDelayLabel = new JLabel("Frame delay:");
        frameDelaySpinner = new JSpinner(new SpinnerNumberModel(100, 25, 1000, 25));
        frameDelaySpinner.addChangeListener(e -> frameDelayChanged());
        JSeparator separator = new JSeparator(SwingConstants.HORIZONTAL);
        JLabel snakeNumLabel = new JLabel("Player snake:");
        snakeNumSpinner = new JSpinner(new SpinnerNumberModel(1, 1, null, 1));
        snakeNumSpinner.addChangeListener(e -> snakeNumChanged());
        JLabel upCtrlLabel = new JLabel("Up key:");
        JLabel downCtrlLabel = new JLabel("Down key:");
        JLabel leftCtrlLabel = new JLabel("Left key:");
        JLabel rightCtrlLabel = new JLabel("Right key:");

        dirCtrlFields = new JTextField[4];
        for (int i = 0; i < dirCtrlFields.length; i++) {
            dirCtrlFields[i] = new JTextField("");
            dirCtrlFields[i].addKeyListener(new CtrlInput(i));
            dirCtrlFields[i].addFocusListener(new CtrlInputFocus(i));
        }

        String fontFamily = dirCtrlFields[0].getFont().getFamily();
        int fontSize = dirCtrlFields[0].getFont().getSize();
        for (JTextField dirCtrlField : dirCtrlFields) {
            dirCtrlField.setEditable(false);
            dirCtrlField.setFont(new Font(fontFamily, Font.BOLD, fontSize));
        }

        snakeNumChanged(); // Fill out player settings with first player's info

        MenuCard gridBag = new MenuCard();
        gridBag.addInGrid(toMainButton2, 0, 0, 2);
        gridBag.addInGrid(mapWidthLabel, 1, 0);
        gridBag.addInGrid(mapWidthSpinner, 1, 1);
        gridBag.addInGrid(mapHeightLabel, 2, 0);
        gridBag.addInGrid(mapHeightSpinner, 2, 1);
        gridBag.addInGrid(frameDelayLabel, 3, 0);
        gridBag.addInGrid(frameDelaySpinner, 3, 1);
        gridBag.addInGrid(separator, 4, 0, 2, new Insets(5, 0, 5, 0));
        gridBag.addInGrid(snakeNumLabel, 5, 0);
        gridBag.addInGrid(snakeNumSpinner, 5, 1);
        gridBag.addInGrid(upCtrlLabel, 6, 0);
        gridBag.addInGrid(dirCtrlFields[0], 6, 1);
        gridBag.addInGrid(downCtrlLabel, 7, 0);
        gridBag.addInGrid(dirCtrlFields[1], 7, 1);
        gridBag.addInGrid(leftCtrlLabel, 8, 0);
        gridBag.addInGrid(dirCtrlFields[2], 8, 1);
        gridBag.addInGrid(rightCtrlLabel, 9, 0);
        gridBag.addInGrid(dirCtrlFields[3], 9, 1);

        mainPanel.add(gridBag, "settings card");
    }

    private void toMainMenuCard() {
        cardLayout.show(mainPanel, "main menu card");
        newGameButton.requestFocus();
    }

    private void toNewGameCard() {
        int minDimension = Math.min((int) mapWidthSpinner.getValue(), (int) mapHeightSpinner.getValue());
        maxSnakeCount = ((minDimension - 11) / 6 + 1) * 4;
        // Done to resize the spinners when the number of digits in maxSnakeCount changes
        playerSnakeSpinner.setModel(new SpinnerNumberModel(1, 0, maxSnakeCount, 1));
        botSnakeSpinner.setModel(new SpinnerNumberModel(0, 0, maxSnakeCount, 1));

        cardLayout.show(mainPanel, "new game card");
        startGameButton.requestFocus();
    }

    private void toSettingsCard() {
        cardLayout.show(mainPanel, "settings card");
        toMainButton2.requestFocus();
    }

    private void playerSnakeCountChanged() {
        if ((int) playerSnakeSpinner.getValue() == 0 && (int) botSnakeSpinner.getValue() == 0) {
            botSnakeSpinner.setValue(1);
        } else if ((int) playerSnakeSpinner.getValue() > playerControls.size()) {
            playerSnakeSpinner.setValue(playerControls.size());
        } else if ((int) playerSnakeSpinner.getValue() + (int) botSnakeSpinner.getValue() > maxSnakeCount) {
            botSnakeSpinner.setValue(maxSnakeCount - (int) playerSnakeSpinner.getValue());
        }
    }

    private void botSnakeCountChanged() {
        if ((int) botSnakeSpinner.getValue() == 0 && (int) playerSnakeSpinner.getValue() == 0) {
            playerSnakeSpinner.setValue(1);
        } else if ((int) botSnakeSpinner.getValue() + (int) playerSnakeSpinner.getValue() > maxSnakeCount) {
            playerSnakeSpinner.setValue(maxSnakeCount - (int) botSnakeSpinner.getValue());
        }
    }

    private void snakeNumChanged() {
        int playerNum = (int) snakeNumSpinner.getValue();
        if (playerNum > playerControls.size()) {
            snakeNumSpinner.setValue(playerControls.size() + 1);
            // It is expected that valid key codes are positive integers
            playerControlsTemp = new int[]{-1, -1, -1, -1};
            for (JTextField dirCtrlField : dirCtrlFields) {
                dirCtrlField.setText("undefined");
            }
        } else {
            playerControlsTemp = playerControls.get(playerNum - 1).clone();
            for (int i = 0; i < dirCtrlFields.length; i++) {
                dirCtrlFields[i].setText(KeyEvent.getKeyText(playerControlsTemp[i]));
            }
        }
    }

    private void frameDelayChanged() {
        int frameDelay = (int) frameDelaySpinner.getValue();
        if (frameDelay % 25 != 0) {
            frameDelaySpinner.setValue(Math.round((float) frameDelay / 25) * 25);
        }
    }

    private void startGame() {
        GamePanel gamePanel = new GamePanel(owner, (int) frameDelaySpinner.getValue(), 10);

        gamePanel.initBoard(
            (int) mapWidthSpinner.getValue(),
            (int) mapHeightSpinner.getValue(),
            wallCheckBox.isSelected()
        );

        for (int i = 0; i < (int) playerSnakeSpinner.getValue(); i++) {
            gamePanel.createPlayerSnake(
                Color.blue,
                playerControls.get(i)[0], playerControls.get(i)[1], playerControls.get(i)[2], playerControls.get(i)[3]
            );
        }
        for (int i = 0; i < (int) botSnakeSpinner.getValue(); i++) {
            gamePanel.createBotSnake(Color.black);
        }

        gamePanel.initLoop();
        owner.changePanel(gamePanel.mainPanel);
    }

    private class CtrlInput extends KeyAdapter {

        int dirIndex;

        private CtrlInput(int setDirIndex) {
            dirIndex = setDirIndex;
        }

        @Override
        public void keyPressed(KeyEvent e) {
            int key = e.getKeyCode();
            if (key == KeyEvent.VK_ESCAPE || key == KeyEvent.VK_SPACE) return;
            if (IntStream.of(playerControlsTemp).anyMatch(x -> x == key)) return;

            playerControlsTemp[dirIndex] = key;
            dirCtrlFields[dirIndex].setText(" " + KeyEvent.getKeyText(key));

            int playerNum = (int) snakeNumSpinner.getValue();
            if (playerNum > playerControls.size()) {
                if (IntStream.of(playerControlsTemp).noneMatch(x -> x == -1)) {
                    playerControls.add(playerControlsTemp.clone());
                }
            } else {
                playerControls.get(playerNum - 1)[dirIndex] = key;
            }
        }
    }

    // Class for giving visual indication of when JTextField is focused on (CtrlInput class helps with this)
    private class CtrlInputFocus implements FocusListener {

        int dirIndex;

        private CtrlInputFocus(int setDirIndex) {
            dirIndex = setDirIndex;
        }

        @Override
        public void focusGained(FocusEvent e) {
            dirCtrlFields[dirIndex].setText(" " + dirCtrlFields[dirIndex].getText().trim());
        }

        @Override
        public void focusLost(FocusEvent e) {
            dirCtrlFields[dirIndex].setText(dirCtrlFields[dirIndex].getText().trim());
        }
    }
}

File: src/com/JavaSnakes/states/menu/MenuPanel.kt
package com.JavaSnakes.states.menu

import com.JavaSnakes.Main
import com.JavaSnakes.states.game.GamePanel
import com.JavaSnakes.states.game.InitSnakes
import com.JavaSnakes.util.CustomJCheckBox
import com.JavaSnakes.util.CustomJSpinner
import com.JavaSnakes.util.GameState
import com.JavaSnakes.util.GridPos
import com.JavaSnakes.util.MenuCard

import java.awt.CardLayout
import java.awt.Dimension
import java.awt.Insets
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JSeparator
import javax.swing.SpinnerNumberModel
import javax.swing.SwingConstants
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.system.exitProcess

class MenuPanel(owner: Main) : GameState(owner) {
    private val cardLayout: CardLayout
    companion object {
        private const val MAIN_MENU_CARD = "main menu card"
        private const val NEW_GAME_CARD = "new game card"
        private const val SETTINGS_CARD = "settings card"
    }

    private val newGameButton = JButton("New Game")
    private val settingsButton = JButton("Settings")
    private val quitButton = JButton("Quit")

    private val startGameButton = JButton("Start Game")
    private val playerCount = CustomJSpinner(1, 0, 1)
    private val botCount = CustomJSpinner(0, 0, 0)
    private val wallsEnabled = CustomJCheckBox()
    private val toMainButton1 = JButton("Back")

    private val toMainButton2 = JButton("Back")
    private val mapW = CustomJSpinner(30, 20, 100)
    private val mapH = CustomJSpinner(20, 20, 100)
    private val frameDelay = CustomJSpinner(100, 25, 1000, 25)
    private val foodCount = CustomJSpinner(2, 1, 10000)
    private val playerSettings = PlayerSettings()

    private val maxSnakeCount: Int
        get() = ((min(mapW.value, mapH.value) - 11) / 6 + 1) * 4

    init {
        mainPanel.layout = CardLayout()

        createMainMenuCard()
        createNewGameCard()
        createSettingsCard()

        cardLayout = mainPanel.layout as CardLayout
        mainPanel.preferredSize = Dimension(300, 300)
    }

    override fun cleanUp() {
        toMainMenuCard()
    }

    private fun createMainMenuCard() {
        newGameButton.addActionListener { toNewGameCard() }
        settingsButton.addActionListener { toSettingsCard() }
        quitButton.addActionListener { exitProcess(0) }

        val gridBag = MenuCard()
        gridBag.addInGrid(newGameButton, 0, 0)
        gridBag.addInGrid(settingsButton, 1, 0)
        gridBag.addInGrid(quitButton, 2, 0)

        mainPanel.add(gridBag, MAIN_MENU_CARD)
    }

    private fun createNewGameCard() {
        startGameButton.addActionListener { startGame() }
        val playerSnakeCountLabel = JLabel("Player count:")
        playerCount.spinner.addChangeListener { playerSnakeCountChanged() }
        val botSnakeCountLabel = JLabel("Bot count:")
        botCount.spinner.addChangeListener { botSnakeCountChanged() }
        wallsEnabled.checkBox.text = "Include walls"
        toMainButton1.addActionListener { toMainMenuCard() }

        val gridBag = MenuCard()
        gridBag.addInGrid(startGameButton, 0, 0, 2)
        gridBag.addInGrid(playerSnakeCountLabel, 1, 0)
        gridBag.addInGrid(playerCount.spinner, 1, 1, padding = Insets(0, 5, 0, 0))
        gridBag.addInGrid(botSnakeCountLabel, 2, 0)
        gridBag.addInGrid(botCount.spinner, 2, 1, padding = Insets(0, 5, 0, 0))
        gridBag.addInGrid(wallsEnabled.checkBox, 3, 0, 2)
        gridBag.addInGrid(toMainButton1, 4, 0, 2)

        mainPanel.add(gridBag, NEW_GAME_CARD)
    }

    private fun createSettingsCard() {
        toMainButton2.addActionListener { toMainMenuCard() }
        val mapWidthLabel = JLabel("Map width:")
        val mapHeightLabel = JLabel("Map height:")
        val frameDelayLabel = JLabel("Frame delay:")
        frameDelay.spinner.addChangeListener { frameDelayChanged() }
        val foodCountLabel = JLabel("Food count:")
        val separator = JSeparator(SwingConstants.HORIZONTAL)
        val snakeNumLabel = JLabel("Player snake:")
        val upCtrlLabel = JLabel("Up key:")
        val downCtrlLabel = JLabel("Down key:")
        val leftCtrlLabel = JLabel("Left key:")
        val rightCtrlLabel = JLabel("Right key:")

        val gridBag = MenuCard()
        gridBag.addInGrid(toMainButton2, 0, 0, 2)
        gridBag.addInGrid(mapWidthLabel, 1, 0)
        gridBag.addInGrid(mapW.spinner, 1, 1)
        gridBag.addInGrid(mapHeightLabel, 2, 0)
        gridBag.addInGrid(mapH.spinner, 2, 1)
        gridBag.addInGrid(frameDelayLabel, 3, 0)
        gridBag.addInGrid(frameDelay.spinner, 3, 1)
        gridBag.addInGrid(foodCountLabel, 4, 0)
        gridBag.addInGrid(foodCount.spinner, 4, 1)
        gridBag.addInGrid(separator, 5, 0, 2, Insets(5, 0, 5, 0))
        gridBag.addInGrid(snakeNumLabel, 6, 0)
        gridBag.addInGrid(playerSettings.snakeNum.spinner, 6, 1)
        gridBag.addInGrid(upCtrlLabel, 7, 0)
        gridBag.addInGrid(playerSettings.dirCtrlFields[0], 7, 1)
        gridBag.addInGrid(downCtrlLabel, 8, 0)
        gridBag.addInGrid(playerSettings.dirCtrlFields[1], 8, 1)
        gridBag.addInGrid(leftCtrlLabel, 9, 0)
        gridBag.addInGrid(playerSettings.dirCtrlFields[2], 9, 1)
        gridBag.addInGrid(rightCtrlLabel, 10, 0)
        gridBag.addInGrid(playerSettings.dirCtrlFields[3], 10, 1)

        mainPanel.add(gridBag, SETTINGS_CARD)
    }

    private fun toMainMenuCard() {
        cardLayout.show(mainPanel, MAIN_MENU_CARD)
        newGameButton.requestFocus()
    }

    private fun toNewGameCard() {
        // Done to resize the spinners when the number of digits in maxSnakeCount changes
        playerCount.setModel(1, 0, maxSnakeCount)
        botCount.setModel(0, 0, maxSnakeCount)

        cardLayout.show(mainPanel, NEW_GAME_CARD)
        startGameButton.requestFocus()
    }

    private fun toSettingsCard() {
        cardLayout.show(mainPanel, SETTINGS_CARD)
        toMainButton2.requestFocus()
    }

    private fun playerSnakeCountChanged() {
        if (playerCount.value == 0 && botCount.value == 0) {
            botCount.value = 1
        } else if (playerCount.value > playerSettings.players.size) {
            playerCount.value = playerSettings.players.size
        } else if (playerCount.value + botCount.value > maxSnakeCount) {
            botCount.value = maxSnakeCount - playerCount.value
        }
    }

    private fun botSnakeCountChanged() {
        if (botCount.value == 0 && playerCount.value == 0) {
            playerCount.value = 1
        } else if (botCount.value + playerCount.value > maxSnakeCount) {
            playerCount.value = maxSnakeCount - botCount.value
        }
    }

    private fun frameDelayChanged() {
        val stepSize = (frameDelay.spinner.model as SpinnerNumberModel).stepSize as Int
        if (frameDelay.value % stepSize != 0) {
            // Value rounded to nearest step size multiple
            frameDelay.value = (frameDelay.value.toFloat() / stepSize).roundToInt() * stepSize
        }
    }

    private fun startGame() {
        GridPos.boardW = mapW.value
        GridPos.boardH = mapH.value

        val initSnakes = InitSnakes(mapW.value, mapH.value)
        for (i in 0 until playerCount.value) {
            initSnakes.addPlayerSnake(playerSettings.players[i])
        }
        initSnakes.addBotSnakes(botCount.value)

        val gamePanel = GamePanel(
            owner, frameDelay.value, 10,
            mapW.value, mapH.value, wallsEnabled.value, foodCount.value, initSnakes.snakes
        )
        owner.toNewPanel(gamePanel)
    }
}
